zschema/keys.py

```diff
@@ -30,6 +30,62 @@ def __cmp__(self, other):
             return cmp(int(self.port), int(other.port))
 
 
+def left_merge(lhs, rhs):
+    """
+    Create a dict with all the keys / values of rhs and lhs, where
+    from lhs are used if a key is in both.
+    """
+    ret = rhs.copy()
+    ret.update(lhs)
+    return ret
+
+
+class TypeFactoryFactory(object):
+    """
+    Factory for TypeFactorys (e.g. SubRecordType is a TypeFactory for
+    SubRecords).
+    """
+
+    def __init__(self, cls, args=None, kwargs=None):
+        """
+        This factory acts as a constructor for cls with the first
+        len(args) positional arguments fixed by args, and with kwargs
+        serving as the default keyword arguments.
+        Note: the positional args can be set here, or in the later call,
+        but not in both.
+        """
+        if not callable(cls):
+            raise TypeError("cls must be callable.")
+
+        if args and not isinstance(args, (list, tuple)):
+            raise TypeError("If present, args must be a list or a tuple.")
+
+        if kwargs and not isinstance(kwargs, dict):
+            raise TypeError("If present, kwargs must be a dict.")
+
+        if args is None:
+            args = []
+
+        if kwargs is None:
+            kwargs = {}
+
+        self.args = tuple(args)
+        self.kwargs = kwargs
+        self.cls = cls
+
+    def __call__(self, *args, **kwargs):
+        """
+        Returns a TypeFactory that returns type instances of cls, using
+        the args positional arguments, and using kwargs as the default
+        values for any keyword arguments.
+        """
+        if len(args) > 0 and len(self.args) > 0:
```

Comment:
This isn't worth blocking over, but FYI, can be written as `if args and self.args`

Reply:
I felt like explicitly using the length made the intention clearer, but maybe that's just because I have a visceral aversion to the idea that an empty list is false.